client/lib/wp/sync-handler/index.js

```diff
@@ -121,26 +163,31 @@ export class SyncHandler {
 			 * WP.com server-side response.
 			 *
 			 * @param {Object} serverResponse - server response object
+			 * @returns {Promise} - promise that resolves with server response object
 			 */
 			const cacheResponse = serverResponse => {
-				// get response object without _headers property
-				let responseWithoutHeaders = Object.assign( {}, serverResponse );
-				delete responseWithoutHeaders._headers;
-
-				let storingData = {
-					__sync: {
-						key,
-						synced: Date.now(),
-						syncing: false
-					},
-					body: responseWithoutHeaders,
-					params: reqParams
-				};
-
-				// add/override gotten data from server-side
-				this.storeRecord( key, storingData ).catch( err => {
-					// @TODO error handling
-					warn( err );
+				return new Promise( ( resolve ) => {
+					// get response object without _headers property
+					let responseWithoutHeaders = Object.assign( {}, serverResponse );
+					delete responseWithoutHeaders._headers;
+
+					let storingData = {
+						__sync: {
+							key,
+							synced: Date.now(),
+							syncing: false
+						},
+						body: responseWithoutHeaders,
+						params: reqParams
+					};
+
+					// add/override gotten data from server-side
+					this.storeRecord( key, storingData ).catch( err => {
+						// @TODO error handling
+						warn( err );
+					} );
+
+					resolve( serverResponse );
```

Comment:
One thing that I've noticed is that in the `checkPagination` function we will modify the local data, in this case the posts-list records will be removed if the pagination changes.
Under these conditions we should avoid storing local data (calling cacheResponse). So all methods that have to process data between the request and the response should be between `networkFetch` and `cacheResponse` calls. In this was we could avoid the final step of this process.

```es6
this.retrieveRecord( key )
  .then( localResponseHandler, recordErrorHandler )
  .then( networkFetch )
  .then( checkPagination, networkErrorHandler )
  .then( cacheResponse );
```